### PR TITLE
Transaction: segregated witness (alternative) (feedback requested)

### PR DESCRIPTION
**Pull request by afk11**
This continues @dcousens work, with some modifications to internals that should set us up well for signing, and some minor tests.

I've made some test vectors out of examples in BIP-1430. I decided to skip adding the OP_CODESEPARATOR cases, because they require special handling, and the code for that doesn't belong in this PR. 

It could certainly be squashed, but wanted to keep the commits separate for now.

### CODE REVIEW

**Comment by afk11** [comment]
Rebased

File: test/transaction.js [hunk 1]
```diff
@@ -233,4 +233,51 @@ describe('Transaction', function () {
...
+        new Buffer('00', 'hex'),
+        new Buffer('51', 'hex')
+      ]
+      tx.setWitness(0, witness)
+      assert.strictEqual(witness, tx.ins[0].witness)
+    })
+  })
+
+  describe('addWitnessInput', function () {
+    it('should add a witness bearing input', function () {
```
Comment (dcousens) [inline test/transaction.js]: what is a witness 'bearing' input?

**Comment by dcousens** [comment]
I think I'll aggregate all the test fixtures first,  so we can be sure about the use cases we need to support,  then put them in a failing PR which we can then merge,  and then resolve with this PR :+1: .

File: src/transaction.js [hunk 2]
```diff
@@ -261,17 +403,33 @@ Transaction.prototype.getId = function () {
   return bufferReverse(this.getHash()).toString('hex')
 }
 
-Transaction.prototype.toBuffer = function (buffer, initialOffset) {
+var SerializationFlags = {}
+SerializationFlags.NONE = 0
+SerializationFlags.ALLOW_SEGWIT = 1 << 1
```
Comment (dcousens) [inline src/transaction.js]: What is the purpose of these?

File: test/transaction.js (hunk 1 above)
Comment (afk11) [inline test/transaction.js]: An input where the witness vector is not empty

File: src/transaction.js (hunk 2 above)
Comment (dcousens) [inline src/transaction.js]: I'd rather no flags at this point

**Comment by afk11** [comment]
Closing in favor of #684

**Pull request closed without merging**